File: src/test/java/no/nav/aareg/teknisk/historikk/KonsumentLoggingTest.java
package no.nav.aareg.teknisk.historikk;

import io.micrometer.core.instrument.MeterRegistry;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.test.web.servlet.client.RestTestClient;

import static no.nav.aareg.teknisk.historikk.KonsumentLogging.AAREG_TEKNISK_HISTORIKK_OPPSLAG_NAVN;
import static no.nav.aareg.teknisk.historikk.KonsumentLogging.DATABEHANDLER_METRIC_TAG_NAVN;
import static no.nav.aareg.teknisk.historikk.KonsumentLogging.OPPSLAGSTYPE_METRIC_TAG_NAVN;
import static no.nav.aareg.teknisk.historikk.KonsumentLogging.ORG_NUMMER_METRIC_TAG_NAVN;
import static no.nav.aareg.teknisk.historikk.api.ApiTestData.TEKNISK_HISTORIKK_RESPONSE_1;
import static no.nav.aareg.teknisk.historikk.api.ApiTestData.gyldigSoekeparameter;
import static org.junit.jupiter.api.Assertions.assertEquals;

class KonsumentLoggingTest extends AaregTekniskHistorikkTest {

    private static final String ENDEPUNKT_URI = "/api/v1/arbeidsforhold";
    private static final String FINN_TEKNISK_HISTORIKK_NAVN = "TekniskHistorikkForArbeidstaker";


    @Autowired
    private RestTestClient restTestClient;

    @Autowired
    private MeterRegistry meterRegistry;

    @Test
    void loggerOppslagsmetrikkerForKonsumenter() {
        aaregServicesStub.stubLegacyTekniskHistorikk(200, TEKNISK_HISTORIKK_RESPONSE_1);

        double forrigeTeller = meterRegistry.counter(
                AAREG_TEKNISK_HISTORIKK_OPPSLAG_NAVN,
                OPPSLAGSTYPE_METRIC_TAG_NAVN, FINN_TEKNISK_HISTORIKK_NAVN,
                ORG_NUMMER_METRIC_TAG_NAVN, TESTORG,
                DATABEHANDLER_METRIC_TAG_NAVN, ""
        ).count();

        restTestClient.post()
                .uri(ENDEPUNKT_URI)
                .headers(headers -> headers.putAll(headerMedAutentisering()))
                .body(gyldigSoekeparameter())
                .exchange()
                .expectStatus().isOk();

        assertEquals(1.0, meterRegistry.counter(
                AAREG_TEKNISK_HISTORIKK_OPPSLAG_NAVN,
                OPPSLAGSTYPE_METRIC_TAG_NAVN, FINN_TEKNISK_HISTORIKK_NAVN,
                ORG_NUMMER_METRIC_TAG_NAVN, TESTORG,
                DATABEHANDLER_METRIC_TAG_NAVN, ""
        ).count() - forrigeTeller);
    }

    @Test
    void loggerOppslagsmetrikkerForDatabehandlere() {
        aaregServicesStub.stubLegacyTekniskHistorikk(200, TEKNISK_HISTORIKK_RESPONSE_1);

        double forrigeTeller = meterRegistry.counter(
                AAREG_TEKNISK_HISTORIKK_OPPSLAG_NAVN,
                OPPSLAGSTYPE_METRIC_TAG_NAVN, FINN_TEKNISK_HISTORIKK_NAVN,
                ORG_NUMMER_METRIC_TAG_NAVN, TESTORG,
                DATABEHANDLER_METRIC_TAG_NAVN, TESTSUPPLIER
        ).count();

        restTestClient.post()
                .uri(ENDEPUNKT_URI)
                .headers(headers -> headers.putAll(headerMedAutentiseringMedDatabehandler()))
                .body(gyldigSoekeparameter())
                .exchange()
                .expectStatus().isOk();

        assertEquals(1.0, meterRegistry.counter(
                AAREG_TEKNISK_HISTORIKK_OPPSLAG_NAVN,
                OPPSLAGSTYPE_METRIC_TAG_NAVN, FINN_TEKNISK_HISTORIKK_NAVN,
                ORG_NUMMER_METRIC_TAG_NAVN, TESTORG,
                DATABEHANDLER_METRIC_TAG_NAVN, TESTSUPPLIER
        ).count() - forrigeTeller);
    }
}

File: src/test/kotlin/no/nav/aareg/teknisk_historikk/KonsumentLoggingTest.kt
package no.nav.aareg.teknisk_historikk

import ch.qos.logback.classic.spi.ILoggingEvent
import ch.qos.logback.core.read.ListAppender
import com.github.tomakehurst.wiremock.client.WireMock
import com.github.tomakehurst.wiremock.junit5.WireMockRuntimeInfo
import com.github.tomakehurst.wiremock.junit5.WireMockTest
import io.micrometer.core.instrument.MeterRegistry
import no.nav.aareg.teknisk_historikk.api.FINN_TEKNISK_HISTORIKK_NAVN
import no.nav.aareg.teknisk_historikk.api.arbeidsforhold1
import no.nav.aareg.teknisk_historikk.api.gyldigSoekeparameter
import no.nav.aareg.teknisk_historikk.models.FinnTekniskHistorikkForArbeidstaker200Response
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.http.HttpEntity

@WireMockTest(httpPort = WIREMOCK_PORT)
class KonsumentLoggingTest : AaregTekniskHistorikkTest() {
    lateinit var logWatcher: ListAppender<ILoggingEvent>

    @Autowired
    lateinit var testRestTemplate: TestRestTemplate

    @Autowired
    lateinit var meterRegistry: MeterRegistry

    companion object {
        val ENDEPUNKT_URI = "/api/v1/arbeidsforhold"
        val AAREG_SERVICES_URI = "/api/beta/tekniskhistorikk"

        @JvmStatic
        @BeforeAll
        internal fun setup(wmRuntimeInfo: WireMockRuntimeInfo) {
            System.setProperty("wiremock.server.baseUrl", wmRuntimeInfo.httpBaseUrl)
        }
    }

    @BeforeEach
    fun setup(wmRuntimeInfo: WireMockRuntimeInfo) {
        logWatcher = ListAppender<ILoggingEvent>().apply { this.start() }
        WireMock.stubFor(
            WireMock.post("/token").willReturn(
                WireMock.okJson("{\"access_token\":\"$testAzureToken\", \"expires_in\": 10}")
            )
        )

        WireMock.stubFor(
            WireMock.get("/wellknown").willReturn(
                WireMock.okJson("{\"token_endpoint\":\"${wmRuntimeInfo.httpBaseUrl}/token\"}")
            )
        )
    }

    @Test
    fun `logger oppslagsmetrikker for konsumenter`(wmRuntimeInfo: WireMockRuntimeInfo) {
        Mockito.`when`(jwtDecoder.decode(testToken)).thenReturn(testJwt)
        WireMock.stubFor(
            WireMock.get(AAREG_SERVICES_URI)
                .willReturn(WireMock.okJson(arbeidsforhold1))
        )

        // TODO nullstill teller før test kjøres
        val forrigeTeller = meterRegistry.counter(AAREG_ARBEIDSFORHOLD_OPPSLAG_NAVN,
            OPPSLAGSTYPE_METRIC_TAG_NAVN, FINN_TEKNISK_HISTORIKK_NAVN,
            ORG_NUMMER_METRIC_TAG_NAVN, testorg,
            DATABEHANDLER_METRIC_TAG_NAVN, "").count()

        testRestTemplate.postForEntity(
            ENDEPUNKT_URI,
            HttpEntity(gyldigSoekeparameter(), headerMedAutentisering()),
            FinnTekniskHistorikkForArbeidstaker200Response::class.java
        )

        assertEquals(1.0, meterRegistry.counter(AAREG_ARBEIDSFORHOLD_OPPSLAG_NAVN,
            OPPSLAGSTYPE_METRIC_TAG_NAVN, FINN_TEKNISK_HISTORIKK_NAVN,
            ORG_NUMMER_METRIC_TAG_NAVN, testorg,
            DATABEHANDLER_METRIC_TAG_NAVN, "").count() - forrigeTeller)
    }

    @Test
    fun `logger oppslagsmetrikker for databehandlere`(wmRuntimeInfo: WireMockRuntimeInfo) {
        Mockito.`when`(jwtDecoder.decode(testToken)).thenReturn(testJwtMedDatabehandler)
        WireMock.stubFor(
            WireMock.get(AAREG_SERVICES_URI)
                .willReturn(WireMock.okJson(arbeidsforhold1))
        )

        // TODO nullstill teller før test kjøres
        val forrigeTeller = meterRegistry.counter(AAREG_ARBEIDSFORHOLD_OPPSLAG_NAVN,
            OPPSLAGSTYPE_METRIC_TAG_NAVN, FINN_TEKNISK_HISTORIKK_NAVN,
            ORG_NUMMER_METRIC_TAG_NAVN, testorg,
            DATABEHANDLER_METRIC_TAG_NAVN, testsupplier).count()

        testRestTemplate.postForEntity(
            ENDEPUNKT_URI,
            HttpEntity(gyldigSoekeparameter(), headerMedAutentisering()),
            FinnTekniskHistorikkForArbeidstaker200Response::class.java
        )

        assertEquals(1.0, meterRegistry.counter(AAREG_ARBEIDSFORHOLD_OPPSLAG_NAVN,
            OPPSLAGSTYPE_METRIC_TAG_NAVN, FINN_TEKNISK_HISTORIKK_NAVN,
            ORG_NUMMER_METRIC_TAG_NAVN, testorg,
            DATABEHANDLER_METRIC_TAG_NAVN, testsupplier).count() - forrigeTeller)
    }
}